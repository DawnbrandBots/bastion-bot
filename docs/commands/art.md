# `/search` command

Diosplay the art for a card! _This command is in development._

## Definitions

[**Password**](https://yugipedia.com/wiki/Password): the number printed in the bottom-left corner of a _Yu-Gi-Oh!_ card.

[**Konami ID**](https://yugipedia.com/wiki/List_of_cards_by_Konami_index_number_(4007%E2%80%935000)): the ID assigned to the card in the [official card database](https://www.db.yugioh-card.com/).

## Parameters

Name | Required? | Description | Type
--- | --- | --- | ---
`input` | ✔ | The password, Konami ID, or name you're searching by. | text
`input-language` | ❌ | The language to search in. | one of the [supported locales](./locale.md#parameters)
`type` | ❌ | Whether you're searching by password, Konami ID, or name. | one of "password", "kid", "name"

## Current behaviour

If `input-language` is unspecified, it is assumed to be the setting for the
channel or server per [Bastion's locale setting](./locale.md).

If `type` is unspecified, infers the type of `input` using these rules:

- numbers are treated as passwords
- numbers that start with `#` are treated as Konami IDs
- anything else is treated as a name

In the first two cases, a direct lookup of the card is performed using that password or Konami ID.
In the latter case, an fuzzy search is performed with the provided `input` in the `input-language`.

The public reply will either be a no-match message or a URL for an image of the card artwork, with the expectation that Discord automatically embeds this.

### Limitations

- Availability and quality of the artwork is consistent.
- Currently no good handling for alternate artworks.

## Next steps

In no particular order:

- fix issues described in limitations
- handle alternative artworks
- depending on how alternative artworks are handled, add a button to display those too
