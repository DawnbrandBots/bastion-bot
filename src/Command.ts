import { RESTPostAPIApplicationCommandsJSONBody } from "discord-api-types/v10";
import { AutocompleteInteraction, ChatInputCommandInteraction } from "discord.js";
import { Logger } from "./logger";
import { Metrics } from "./metrics";
import { serialiseInteraction } from "./utils";

export abstract class Command {
	static get meta(): RESTPostAPIApplicationCommandsJSONBody {
		throw new Error("Not implemented");
	}

	constructor(protected metrics: Metrics) {}

	get meta(): RESTPostAPIApplicationCommandsJSONBody {
		return (this.constructor as typeof Command).meta;
	}

	protected abstract get logger(): Logger;

	/**
	 * Execute this command in response to a Slash Command. May throw exceptions,
	 * which will be captured and logged appropriately, and feedback will be
	 * provided to the user.
	 *
	 * @param interaction
	 * @returns latency metric in milliseconds
	 */
	protected abstract execute(interaction: ChatInputCommandInteraction): Promise<number>;

	/**
	 * Run this command in response to user interaction from start to finish.
	 * Does not throw exceptions.
	 *
	 * @param interaction
	 */
	async run(interaction: ChatInputCommandInteraction): Promise<void> {
		try {
			this.logger.verbose(
				serialiseInteraction(interaction, { event: "attempt", ping: interaction.client.ws.ping })
			);
			const latency = await this.execute(interaction);
			this.logger.verbose(serialiseInteraction(interaction, { event: "success", latency }));
			this.metrics.writeCommand(interaction, latency);
		} catch (error) {
			this.metrics.writeCommand(interaction, -1);
			this.logger.error(serialiseInteraction(interaction), error);
			let method;
			if (interaction.replied) {
				method = "followUp" as const;
			} else if (interaction.deferred) {
				method = "editReply" as const;
			} else {
				method = "reply" as const;
			}
			await interaction[method]("Something went wrong. Please try again later.").catch(e =>
				this.logger.error(serialiseInteraction(interaction), e)
			);
		}
	}
}

export abstract class AutocompletableCommand extends Command {
	abstract autocomplete(interaction: AutocompleteInteraction): Promise<void>;
}
