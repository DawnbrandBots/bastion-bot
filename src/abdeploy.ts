import { Snowflake } from "discord.js";
import { readFile } from "fs/promises";
import { inject, singleton } from "tsyringe";
import { getLogger } from "./logger";

@singleton()
export class ABDeploy {
	#logger = getLogger("abdeploy");

	protected interval: NodeJS.Timeout;
	protected registry: Set<Snowflake> | null = null;

	constructor(@inject("abdeployJson") private file: string) {
		// Promise can resolve after constructor, just don't until next trigger to populate
		this.update(true);
		this.interval = setInterval(() => this.update(), 60000).unref();
	}

	protected async update(initial = false): Promise<void> {
		// Synchronise on the hour
		if (initial || new Date().getMinutes() === 0) {
			this.#logger.info(`Updating from ${this.file}`);
			try {
				const json = await readFile(this.file, { encoding: "utf8" });
				this.registry = new Set(JSON.parse(json));
				this.#logger.info(`Read ${this.registry.size} entries`);
			} catch (error) {
				// Info level because file may not exist and this is okay
				this.#logger.info(error);
			}
		}
	}

	has(server: Snowflake): boolean {
		return !!this.registry?.has(server);
	}

	/**
	 * Use to clean up this object, after which it should not be used.
	 */
	finalise(): void {
		clearInterval(this.interval);
		this.registry = null;
	}
}
