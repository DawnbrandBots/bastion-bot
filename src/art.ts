import { Static } from "@sinclair/typebox";
import { Debug } from "debug";
import {
	ActionRowBuilder,
	BaseMessageOptions,
	ButtonBuilder,
	ButtonInteraction,
	ButtonStyle,
	ChatInputCommandInteraction,
	ComponentType,
	DiscordAPIError,
	DiscordjsErrorCodes,
	Message
} from "discord.js";
import { Got } from "got";
import { t, useLocale } from "ttag";
import { yugipediaFileRedirect } from "./card";
import { CardSchema } from "./definitions";
import { RushCardSchema } from "./definitions/rush";
import { Locale } from "./locale";
import { Logger, getLogger } from "./logger";
import { serialiseInteraction } from "./utils";

export async function checkYugipediaRedirect(got: Got, url: string, warn: Debug["log"]): Promise<boolean> {
	try {
		const response = await got(url, {
			method: "HEAD",
			followRedirect: false,
			timeout: 2000
		});
		if (response.statusCode === 302) {
			// MediaWiki Special:Redirect/file should only use 302s
			return true;
		} else if (response.statusCode !== 404) {
			warn({ redirectStatusCode: response.statusCode });
		}
	} catch (error) {
		warn(error);
	}
	return false;
}

export class ArtSwitcher {
	private readonly logger: Logger;

	private readonly labelButton = new ButtonBuilder()
		.setCustomId("label")
		.setStyle(ButtonStyle.Primary)
		.setDisabled(true);
	private readonly prevButton = new ButtonBuilder()
		.setCustomId("prev")
		.setEmoji("⬅")
		.setStyle(ButtonStyle.Secondary)
		.setDisabled(true);
	private readonly nextButton = new ButtonBuilder()
		.setCustomId("next")
		.setEmoji("➡")
		.setStyle(ButtonStyle.Secondary)
		.setDisabled(true);
	private readonly components = [
		new ActionRowBuilder<ButtonBuilder>().addComponents(this.labelButton, this.prevButton, this.nextButton)
	];

	private index = 0;

	constructor(
		private readonly images: NonNullable<Static<typeof RushCardSchema | typeof CardSchema>["images"]>,
		context: string
	) {
		this.labelButton.setLabel(this.label);
		this.nextButton.setDisabled(images.length === 1);
		this.logger = getLogger(`command:${context}:switcher`);
	}

	private get label(): string {
		return `${this.index + 1} / ${this.images.length}`;
	}

	private get currentImage(): string {
		const illustration = this.images[this.index].illustration;
		if (illustration) {
			return yugipediaFileRedirect(illustration);
		} else {
			const image = this.images[this.index].image;
			return yugipediaFileRedirect(image);
		}
	}

	private get replyOptions(): BaseMessageOptions {
		return {
			content: this.currentImage,
			components: this.components
		};
	}

	private onClick(interaction: ButtonInteraction): void {
		if (interaction.customId === "prev") {
			if (this.index > 0) {
				this.index--;
			}
		} else {
			if (this.index < this.images.length - 1) {
				this.index++;
			}
		}
		this.prevButton.setDisabled(this.index === 0);
		this.nextButton.setDisabled(this.index === this.images.length - 1);
		this.labelButton.setLabel(this.label);
	}

	async send(
		parentInteraction: ChatInputCommandInteraction,
		method: "reply" | "editReply",
		resultLanguage: Locale
	): Promise<Message> {
		const reply = await parentInteraction[method]({ ...this.replyOptions, fetchReply: true });
		// Don't bother setting up button handler if all buttons are disabled
		if (this.images.length === 1) {
			return reply;
		}
		const filter = (childInteraction: ButtonInteraction): boolean => {
			this.logger.info(serialiseInteraction(parentInteraction), `click: ${childInteraction.user.id}`);
			if (childInteraction.user.id === parentInteraction.user.id) {
				return true;
			}
			useLocale(resultLanguage);
			childInteraction
				.reply({
					content: t`Buttons can only be used by the user who called Bastion.`,
					ephemeral: true
				})
				.catch(e => this.logger.error(serialiseInteraction(parentInteraction), e));
			return false;
		};
		// Set up the button handler (don't await) and return the initial reply
		const awaitOptions = { filter, componentType: ComponentType.Button, time: 60000 } as const;
		const then = async (childInteraction: ButtonInteraction): Promise<void> => {
			this.onClick(childInteraction);
			// We have to set up the handler again because Discord.js is forcing the maximum number of events to 1
			// https://github.com/discordjs/discord.js/blob/fb70df817c6566ca100d57ec1878a4573489a43d/packages/discord.js/src/structures/InteractionResponse.js#L30
			(await childInteraction.update(this.replyOptions))
				.awaitMessageComponent(awaitOptions)
				.then(then)
				.catch(catcher);
		};
		const catcher = async (err: DiscordAPIError): Promise<void> => {
			// a rejection can just mean the timeout was reached without a response
			// otherwise, though, we want to treat it as a normal error
			if (err.code !== DiscordjsErrorCodes.InteractionCollectorError) {
				this.logger.error(serialiseInteraction(parentInteraction), err);
			} else {
				this.logger.verbose(serialiseInteraction(parentInteraction), err);
				if (err.message.split("reason: ")[1].toLowerCase().includes("delete")) {
					return;
				}
			}
			// disable original buttons, except when deleted since there's nothing toe dit
			this.prevButton.setDisabled(true);
			this.nextButton.setDisabled(true);
			try {
				await parentInteraction.editReply(this.replyOptions);
			} catch (e) {
				this.logger.error(serialiseInteraction(parentInteraction), e);
			}
		};
		reply.awaitMessageComponent(awaitOptions).then(then).catch(catcher);
		return reply;
	}
}
