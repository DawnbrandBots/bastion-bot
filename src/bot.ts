import { ActivityType, Client, GatewayIntentBits, Options, Partials } from "discord.js";
import { injectAll, injectable } from "tsyringe";
import { Listener } from "./events";
import { getLogger } from "./logger";
import { serialiseServer } from "./utils";

@injectable()
export class BotFactory {
	// This is not type-safe! TypeScript doesn't support declaring an array of generic
	// Listeners without a specialization though. This requires an existential type:
	// https://github.com/microsoft/TypeScript/issues/14466
	// https://stackoverflow.com/questions/65129070/defining-an-array-of-differing-generic-types-in-typescript
	// https://stackoverflow.com/questions/292274/what-is-an-existential-type
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	constructor(@injectAll("Listener") private listeners: Listener<any>[]) {}

	createInstance(): Client {
		Options.createDefault().shardCount;
		const logger = getLogger("events");

		const defaults = Options.createDefault();
		const shard = parseInt(`${process.env.DISCORD_SHARD}`) - 1;
		const shards =
			`${process.env.DISCORD_TOTAL_SHARDS}` === "auto" ? "auto" : isNaN(shard) ? defaults.shards : shard;
		const shardCount = parseInt(`${process.env.DISCORD_TOTAL_SHARDS}`) || defaults.shardCount;

		const bot = new Client({
			intents: [
				GatewayIntentBits.Guilds,
				GatewayIntentBits.GuildMessages,
				GatewayIntentBits.MessageContent,
				// Intents.FLAGS.GUILD_MESSAGE_REACTIONS,
				GatewayIntentBits.DirectMessages
				// Intents.FLAGS.DIRECT_MESSAGE_REACTIONS
			],
			partials: [
				Partials.Channel, // for direct message events
				Partials.Message // for message deletion events
			],
			shards,
			shardCount,
			makeCache: Options.cacheWithLimits({
				GuildEmojiManager: 0,
				GuildTextThreadManager: 0,
				MessageManager: 0 // reduce cache per channel from default 200
			})
		});

		const state = process.env.BOT_PRESENCE || "🔎 <card name> to search!";
		function setActivity(): void {
			bot.user?.setActivity(state, { type: ActivityType.Custom });
		}

		bot.on("warn", message => logger.warn(`Shard ${bot.shard}: ${message}`));
		bot.on("error", message => logger.error(`Shard ${bot.shard}: ${message}`));
		bot.on("shardReady", shard => logger.notify(`Shard ${shard} ready`));
		bot.on("shardReconnecting", shard => logger.info(`Shard ${shard} reconnecting`));
		bot.on("shardResume", (shard, replayed) => logger.info(`Shard ${shard} resumed: ${replayed} events replayed`));
		bot.on("shardDisconnect", (event, shard) => logger.notify(`Shard ${shard} disconnected (${event.code})`));
		bot.on("shardError", (error, shard) => logger.error(`Shard ${shard} error:`, error));
		bot.on("guildCreate", guild => logger.notify(`Guild create: ${serialiseServer(guild)}`));
		bot.on("guildDelete", guild => logger.notify(`Guild delete: ${serialiseServer(guild)}`));
		bot.on("ready", () => {
			logger.notify(`Logged in as ${bot.user?.tag} - ${bot.user?.id}`);
			// May get blanked by reconnections and other things, so reapply every hour
			setActivity();
			setInterval(setActivity, 1000 * 60 * 60);
		});

		bot.once("ready", () => {
			let lastEventTimestamp = Date.now();
			let incidentNotified = false;
			bot.on("messageCreate", () => void (lastEventTimestamp = Date.now()));
			setInterval(() => {
				if (Date.now() - lastEventTimestamp > 60000) {
					if (!incidentNotified) {
						logger.warn("⚠️ No messageCreate gateway events received for one minute ⚠️");
						incidentNotified = true;
					}
				} else {
					if (incidentNotified) {
						logger.notify("Started receiving messageCreate gateway events again");
					}
					incidentNotified = false;
				}
			}, 1000).unref();
			logger.info("Registered messageCreate gateway event frequency early warning canary");
		});

		for (const listener of this.listeners) {
			bot.on(listener.type, (...args) => listener.run(...args));
		}

		return bot;
	}
}
