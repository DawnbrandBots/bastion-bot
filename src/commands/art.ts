import { RESTPostAPIApplicationCommandsJSONBody } from "discord-api-types/v10";
import { ChatInputCommandInteraction } from "discord.js";
import { Got } from "got";
import { inject, injectable } from "tsyringe";
import { c, t, useLocale } from "ttag";
import { Command } from "../Command";
import { ArtSwitcher } from "../art";
import { getCard, getCardSearchOptions, getRubylessCardName, masterDuelIllustration } from "../card";
import {
	LocaleProvider,
	buildLocalisedCommand,
	everywhereCommand,
	getKonamiIdSubcommand,
	getNameSubcommand,
	getPasswordSubcommand
} from "../locale";
import { Logger, getLogger } from "../logger";
import { Metrics } from "../metrics";
import { replyLatency } from "../utils";

@injectable()
export class ArtCommand extends Command {
	#logger = getLogger("command:art");

	constructor(
		metrics: Metrics,
		@inject("LocaleProvider") private locales: LocaleProvider,
		@inject("got") private got: Got
	) {
		super(metrics);
	}

	static override get meta(): RESTPostAPIApplicationCommandsJSONBody {
		const builder = buildLocalisedCommand(
			everywhereCommand(),
			() => c("command-name").t`art`,
			() => c("command-description").t`Display the art for a card!`
		);
		const nameSubcommand = getNameSubcommand(
			() => c("command-option-description").t`Display the art for the card with this name.`
		);
		const passwordSubcommand = getPasswordSubcommand(
			() => c("command-option-description").t`Display the art for the card with this password.`
		);
		const konamiIdSubcommand = getKonamiIdSubcommand(
			() => c("command-option-description").t`Display the art for the card with this official database ID.`
		);
		builder.addSubcommand(nameSubcommand).addSubcommand(passwordSubcommand).addSubcommand(konamiIdSubcommand);
		return builder.toJSON();
	}

	protected override get logger(): Logger {
		return this.#logger;
	}

	protected override async execute(interaction: ChatInputCommandInteraction): Promise<number> {
		const { type, input, resultLanguage, inputLanguage } = await getCardSearchOptions(interaction, this.locales);
		const card = await getCard(this.got, type, input, inputLanguage);
		let reply;
		if (!card) {
			useLocale(resultLanguage);
			reply = await interaction.reply({
				content: t`Could not find a card matching \`${input}\`!`,
				fetchReply: true
			});
		} else if (!card.images) {
			const name = getRubylessCardName(card.name[resultLanguage] || `${card.konami_id}`, resultLanguage);
			useLocale(resultLanguage);
			reply = await interaction.reply({
				content: t`Could not find art for \`${name}\`!`,
				fetchReply: true
			});
		} else {
			// Avoid the latency of checking a wiki redirect on every command
			if (card.master_duel_rarity && !card.images[0].illustration) {
				card.images[0].illustration = masterDuelIllustration(card);
			}
			const switcher = new ArtSwitcher(card.images, "art");
			reply = await switcher.send(interaction, "reply", resultLanguage);
		}
		return replyLatency(reply, interaction);
	}
}
