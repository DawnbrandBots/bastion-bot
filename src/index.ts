import fs from "fs";
import { po } from "gettext-parser";
import os from "os";
import path from "path";
import { container } from "tsyringe";
import { addLocale } from "ttag";
import { BotFactory } from "./bot";
import { Command } from "./Command";
import { classes, registerSlashCommands } from "./commands";
import { InteractionListener, MessageListener } from "./events";
import { LocaleProvider, SQLiteLocaleProvider } from "./locale";
import { getLogger } from "./logger";
import { Metrics } from "./metrics";

if (process.argv.length > 2 && process.argv[2] === "--deploy-slash") {
	// We don't need to verify the bigint typing since this CLI operation will safely fail
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	registerSlashCommands(process.argv[3] as any);
} else {
	const logger = getLogger("index");

	const metricsDb = process.argv[2]
		? path.join(process.argv[2], "metrics.db3")
		: path.join(os.tmpdir(), "bastion-metrics.db3");
	logger.info(`Storing metrics in ${metricsDb}`);
	container.register<string>("metricsDb", {
		useValue: metricsDb
	});

	const localeDb = process.argv[2]
		? path.join(process.argv[2], "locales.db3")
		: path.join(os.tmpdir(), "bastion-locales.db3");
	logger.info(`Storing locales in ${localeDb}`);
	container.register<string>("localeDb", {
		useValue: localeDb
	});

	for (const file of fs.readdirSync("./translations", { withFileTypes: true })) {
		if (file.isFile() && file.name.endsWith(".po")) {
			const jsonpo = po.parse(fs.readFileSync(`./translations/${file.name}`));
			const locale = file.name.split(".po")[0];
			addLocale(locale, jsonpo);
			logger.info(`Loaded translations for locale ${locale}`);
		}
	}

	//container.registerSingleton<Metrics>(Metrics);
	container.registerSingleton<LocaleProvider>("LocaleProvider", SQLiteLocaleProvider);
	classes.forEach(Class => container.register<Command>("Command", { useClass: Class }));
	container.register<InteractionListener>("Listener", { useClass: InteractionListener });
	container.register<MessageListener>("Listener", { useClass: MessageListener });
	//container.register<BotFactory>(BotFactory, { useClass: BotFactory });

	const bot = container.resolve(BotFactory).createInstance();
	process.once("SIGTERM", () => {
		bot.destroy();
		container.resolve(Metrics).destroy();
		container.resolve(SQLiteLocaleProvider).destroy();
	});
	// Implicitly use DISCORD_TOKEN
	bot.login();
}
