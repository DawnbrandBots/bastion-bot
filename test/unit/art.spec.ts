import { ChatInputCommandInteraction, DiscordjsError, DiscordjsErrorCodes } from "discord.js";
import { ArtSwitcher } from "../../src/art";

describe("ArtSwitcher", () => {
	const awaitMessageComponent = jest.fn();
	const editReply = jest.fn();
	const reply = jest.fn();
	const mockInteraction = { editReply, reply };
	const interaction = mockInteraction as unknown as ChatInputCommandInteraction;
	beforeEach(() => {
		awaitMessageComponent.mockReset();
		editReply.mockReset().mockReturnValue({ awaitMessageComponent });
		reply.mockReset().mockReturnValue({ awaitMessageComponent });
	});
	it.each(["reply", "editReply"] as const)("shows single illustration using %s", async method => {
		const switcher = new ArtSwitcher(
			[
				{
					illustration: "MekkKnightCrusadiaAvramax-MADU-EN-VG-artwork.png",
					index: 1,
					image: "MekkKnightCrusadiaAvramax-RA01-EN-SR-1E.png"
				}
			],
			"test"
		);
		await switcher.send(interaction, method, "en");
		expect(mockInteraction[method]).toHaveBeenCalledTimes(1);
		expect(mockInteraction[method].mock.calls[0][0]).toMatchSnapshot();
	});
	it.each(["reply", "editReply"] as const)("shows image if no illustrations using %s", async method => {
		const switcher = new ArtSwitcher(
			[
				{
					index: 1,
					image: "Shuttleroid-PP11-JP-ScR.jpg"
				}
			],
			"test"
		);
		await switcher.send(interaction, method, "en");
		expect(mockInteraction[method]).toHaveBeenCalledTimes(1);
		expect(mockInteraction[method].mock.calls[0][0]).toMatchSnapshot();
	});
	const CARTESIA_IMAGES = [
		{
			illustration: "BlazingCartesiatheVirtuous-MADU-JP-VG-artwork.png",
			index: 1,
			image: "BlazingCartesiatheVirtuous-DABL-JP-SR.png"
		},
		{
			illustration: "BlazingCartesiatheVirtuous-MADU-EN-VG-artwork.png",
			index: "1.1",
			image: "BlazingCartesiatheVirtuous-MP23-EN-PScR-1E.png"
		}
	];
	it.each(["reply", "editReply"] as const)("shows first of two illustrations using %s", async method => {
		awaitMessageComponent.mockRejectedValue(
			Reflect.construct(DiscordjsError, [DiscordjsErrorCodes.InteractionCollectorError])
		);
		const interaction = { ...mockInteraction, user: {} } as unknown as ChatInputCommandInteraction;
		const switcher = new ArtSwitcher(CARTESIA_IMAGES, "test");
		await switcher.send(interaction, method, "en");
		expect(mockInteraction[method]).toHaveBeenCalledTimes(1);
		expect(mockInteraction[method].mock.calls[0][0]).toMatchSnapshot();
	});
	it.todo("shows second of two illustrations");
});
