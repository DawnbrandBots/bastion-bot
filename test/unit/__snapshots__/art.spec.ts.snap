// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtSwitcher shows first of two illustrations using editReply 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "label",
          "disabled": true,
          "emoji": undefined,
          "label": "1 / 2",
          "style": 1,
          "type": 2,
        },
        {
          "custom_id": "prev",
          "disabled": true,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "⬅",
          },
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "next",
          "disabled": false,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "➡",
          },
          "style": 2,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "https://yugipedia.com/wiki/Special:Redirect/file/BlazingCartesiatheVirtuous-MADU-JP-VG-artwork.png?utm_source=bastion",
  "fetchReply": true,
}
`;

exports[`ArtSwitcher shows first of two illustrations using reply 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "label",
          "disabled": true,
          "emoji": undefined,
          "label": "1 / 2",
          "style": 1,
          "type": 2,
        },
        {
          "custom_id": "prev",
          "disabled": true,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "⬅",
          },
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "next",
          "disabled": false,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "➡",
          },
          "style": 2,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "https://yugipedia.com/wiki/Special:Redirect/file/BlazingCartesiatheVirtuous-MADU-JP-VG-artwork.png?utm_source=bastion",
  "fetchReply": true,
}
`;

exports[`ArtSwitcher shows image if no illustrations using editReply 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "label",
          "disabled": true,
          "emoji": undefined,
          "label": "1 / 1",
          "style": 1,
          "type": 2,
        },
        {
          "custom_id": "prev",
          "disabled": true,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "⬅",
          },
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "next",
          "disabled": true,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "➡",
          },
          "style": 2,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "https://yugipedia.com/wiki/Special:Redirect/file/Shuttleroid-PP11-JP-ScR.jpg?utm_source=bastion",
  "fetchReply": true,
}
`;

exports[`ArtSwitcher shows image if no illustrations using reply 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "label",
          "disabled": true,
          "emoji": undefined,
          "label": "1 / 1",
          "style": 1,
          "type": 2,
        },
        {
          "custom_id": "prev",
          "disabled": true,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "⬅",
          },
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "next",
          "disabled": true,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "➡",
          },
          "style": 2,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "https://yugipedia.com/wiki/Special:Redirect/file/Shuttleroid-PP11-JP-ScR.jpg?utm_source=bastion",
  "fetchReply": true,
}
`;

exports[`ArtSwitcher shows single illustration using editReply 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "label",
          "disabled": true,
          "emoji": undefined,
          "label": "1 / 1",
          "style": 1,
          "type": 2,
        },
        {
          "custom_id": "prev",
          "disabled": true,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "⬅",
          },
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "next",
          "disabled": true,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "➡",
          },
          "style": 2,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "https://yugipedia.com/wiki/Special:Redirect/file/MekkKnightCrusadiaAvramax-MADU-EN-VG-artwork.png?utm_source=bastion",
  "fetchReply": true,
}
`;

exports[`ArtSwitcher shows single illustration using reply 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "label",
          "disabled": true,
          "emoji": undefined,
          "label": "1 / 1",
          "style": 1,
          "type": 2,
        },
        {
          "custom_id": "prev",
          "disabled": true,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "⬅",
          },
          "style": 2,
          "type": 2,
        },
        {
          "custom_id": "next",
          "disabled": true,
          "emoji": {
            "animated": false,
            "id": undefined,
            "name": "➡",
          },
          "style": 2,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "https://yugipedia.com/wiki/Special:Redirect/file/MekkKnightCrusadiaAvramax-MADU-EN-VG-artwork.png?utm_source=bastion",
  "fetchReply": true,
}
`;
